### ISSUE TITLE
Add TypeScript support

### ISSUE THREAD

**Issue by sebastianwahn** [sequelize/cli#328]
Since coffeescript is supported, i think it would be cool to have a typescript support also.

**Expectation:**
```run node_modules/sequelize-cli/bin/sequelize model:create --name Collection --attributes id:integer,name:string --typescript``
`

Will produce a model which ends in '.ts' instead of '.js'.

**Justification:**
In projects where typescript is used, javascript files are sometimes ignored (because they are produced through typescript). Therefore, the model will be ignored by default. Afterwards, a developer must edit the model/migration.

**Result:**
Tiny bit of work removed when developing with typescript.

**Comment by pascalbayer** [comment]
+1

**Comment by Vayrex** [comment]
Really good proposition +1

**Comment by felixfbecker** [comment]
The problem I see here is that the pattern I found to be the best for TypeScript is totally different than the approach sequelize-cli uses. The pattern of the exported closure and a static `associate()` method is really hard to type (how do you type the `models` parameter?). Using late imports works much better.

**Comment by kenhowardpdx** [comment]
Does "Add TypeScript support" also include generating migrations? If so the only change there is to save the file as `.ts` instead of `.js`.

**Comment by felixfbecker** [comment]
What does this have to do with classes?

**Comment by sebastianwahn** [comment]
Sorry, you are right. Thought about another thing.
Regarding the TypeScript, it would be sufficient to have the .ts ending and some TypeScript syntax (lambda functions, ...).

Another feature would be to have the classes above instead of the current syntax:
I can define the above class, it will be derived from a sequlize-parent-model (or something).
Afterwards this class is scanned for its properties etc.
This thing would go into the direction of EntityFramework (C#) and Hibernate (Java).

**Comment by felixfbecker** [comment]
You are thinking about models, not migration files. A migration file just exports two functions, up and down.

**Comment by Mikhus** [comment]
+1 to this. Actually it should look not simply as file saved with .ts extension, but contain a slightly different contents in it, for example:

~~~typescript
import { QueryInterface } from 'sequelize';
import { DataType } from 'sequelize-typescript';

export async function up(query: QueryInterface) {
    /*
      Add altering commands here.

      Example:
      await query.createTable('users', { id: DataType.INTEGER });
      console.log('Table users created!');
    */
}

export async function down(query: QueryInterface) {
    /*
      Add reverting commands here.

      Example:
      await query.dropTable('users');
      console.log('Table users dropped!');
    */
}
~~~

So:
 1) it should save with different extension
 2) it should put different contents inside....

Maybe it's even better to provide those possibilities in more generic way, like add an ability to provide user-defined template for those files. An I will be able to construct myself which dialect of javascript I use...

**Comment by kenhowardpdx** [comment]
@Mikhus YES! Native module support is required. The nice thing about your sample code is that you only introduce TypeScript to utilize QueryInterface. 👍

**Comment by foisonocean** [comment]
+`-1

**Comment by nagapavan** [comment]
+1

**Comment by limichange** [comment]
+1

**Comment by GeeWee** [comment]
+1

**Comment by MarcoPasqualini** [comment]
+1

**Comment by virgenherrera** [comment]
+7
:D

**Comment by felixfbecker** [comment]
Please don't +1 issues. It clutters the thread without adding value to the discussion and spams maintainers with notifications. Use GitHub reactions to upvote features.
![image](https://cloud.githubusercontent.com/assets/10532611/18344735/bb60c2f6-75b8-11e6-88a9-822fae298110.png)

**Comment by virgenherrera** [comment]
thks @felixfbecker i'do that.

**Comment by georgeedwards** [comment]
@felixfbecker I have not used it in a project (but am planning to on mynext project), but have you seen the [sequelize-typescript)[https://github.com/RobinBuschmann/sequelize-typescript] project - it looks like a really interesting way of making Sequelize a bit more typescripty...

I am not sure if that makes the generated application prescribe a package that people might not be familiar with / want.

**Comment by ncphillips** [comment]
Another option would be to treat the `models` directory as a package, and instead of generating different code for typescript, just generating the `types.d.ts` declaration files.

This might look something like:

Directory Structure
```
models/
  index.js
  index.d.ts
  package.json
  user.js
  user.d.ts
```
  
package.json
```json
{
  "name": "create-observable-thunk",
  "main": "./index.js",
  "typings": "./index.d.ts",
}
```

types.d.ts
```typescript
import * as Sequelize from "sequelize"
export const Sequelize: Sequelize
export const sequelize: Sequelize.Sequelize
export default namespace db {

}
```

user.d.ts
```typescript
import * as Sequelize from "sequelize"

interface Attributes {
  // ...
}

export default namespace db {
  Cat: Sequelize.Model
}
```

Note I haven't made sure these typings made sense.

**Comment by douglas-treadwell** [comment]
Until sequelize-cli has a TypeScript support option, the people on this thread may want to check out https://github.com/douglas-treadwell/sequelize-cli-typescript.  I forked sequelize-cli and made a version that outputs TypeScripts (including migrations).  See the README section "Differences...", but otherwise it works like regular sequelize-cli and should be a drop-in replacement.

**Comment by abdoolly** [comment]
checkout this package which may help you to make typescript models and ease out model initialization
https://www.npmjs.com/package/ts-sequelize-models

**Comment by dantaub** [comment]
I would request that instead of `--typescript` we would have `--generatedExtension ts` 

If coupled with registering transformations based on extension, we could support output in coffeescript, livescript, es6, etc

**Comment by philippeauriach** [comment]
If it can be of any help, I managed to launch sequelize cli with typescript files using babel's plugin 'preset-typescript' in my `.sequelizerc` file. Not sure if its the best way to do, but it works.

`.sequelizerc`
```
var path = require("path");

require('@babel/register')({
  plugins: [
    '@babel/plugin-transform-runtime',
    'babel-plugin-add-module-exports'
  ],
  presets: [
    '@babel/preset-typescript',
    '@babel/preset-env',
    '@babel/preset-react'
  ],
  extensions: ['.js','.jsx','.ts','.tsx']
});

module.exports = {
  'config': path.resolve('server/db', 'config.js'),
  'migrations-path': path.resolve('server/db', 'migrate'),
  'models-path': path.resolve('server/db', 'models'),
  'seeders-path': path.resolve('server/db', 'seeders')
};
```

**Comment by mankinchi** [comment]
Any updates on this?

**Comment by mrdulin** [comment]
+1

**Comment by mankinchi** [comment]
@mrdulin I ended up using [this](https://github.com/RobinBuschmann/sequelize-typescript)

**Comment by intellix** [comment]
I assume the opinion has changed here since Sequelize 5 exports TypeScript types and could be of some value here. Would love type hints when writing my migrations as it's currently trial and error as the documentation often just says: "object" for options

**Comment by AnthonyLzq** [comment]
It seems that there has been no progress on this requirement, so does anyone knows a reliable way to perform migrations with TS?

**Comment by CyanoFresh** [comment]
would like to know too

**Comment by mjovanc** [comment]
Any news? This seems like a very good proposition.

**Comment by virgenherrera** [comment]
I'd better opt to migrate to Typeorm, since this library looks like it is continuously ignoring Typescript support.

**Comment by WikiRik** [comment]
We're hard at work at improving TypeScript support for the main repo and we'll work on CLI after we've done most of our planned changes. v7 will have much better TypeScript support but we're open to PRs of the community in the meantime

**Comment by IgorArnaut** [comment]
> We're hard at work at improving TypeScript support for the main repo and we'll work on CLI after we've done most of our planned changes. v7 will have much better TypeScript support but we're open to PRs of the community in the meantime
Any updates? It's 2025.

**Comment by virgenherrera** [comment]
> > We're hard at work at improving TypeScript support for the main repo and we'll work on CLI after we've done most of our planned changes. v7 will have much better TypeScript support but we're open to PRs of the community in the meantime
> 
> Any updates? It's 2025.

Since the last release of this ORM was on Nov 3, 2023, I think that no further updates will come to this project

**Comment by nbkhope** [comment]
> Any updates? It's 2025.

Now it's 2026 😉

Have you tried running sequelize cli like this?

```
tsx ./node_modules/sequelize-cli/lib/sequelize db:migrate --config config/config.ts
```

That way, instead of using node, it uses tsx to run the migrations written with the .ts extension
